docker/scripts/entrypoint.sh

```diff
@@ -61,11 +63,50 @@ until /opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain -server
 echo "create the zk root dir for bookkeeper"
 /opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain -server ${BK_zkServers} create ${BK_CLUSTER_ROOT_PATH}
 
-echo "format zk metadata"
-echo "please ignore the failure, if it has already been formatted, "
 export BOOKIE_CONF=/opt/bookkeeper/conf/bk_server.conf
 export SERVICE_PORT=$PORT0
-/opt/bookkeeper/bin/bookkeeper shell metaformat -n || true
+
+# Init the cluster if required znodes not exist in Zookeeper.
+# Use ephemeral zk node as lock to keep initialize atomic.
+/opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain -server ${BK_zkServers} stat ${BK_zkLedgersRootPath}/available/readonly
+if [ $? -eq 0 ]; then
+    echo "Metadata of cluster already exists, no need format"
+else
+    # create ephemeral zk node bkInitLock, initiator who this node, then do init; other initiators will wait.
+    /opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain -server ${BK_zkServers} create -e ${BK_CLUSTER_ROOT_PATH}/bkInitLock
+    if [ $? -eq 0 ]; then
+        # bkInitLock created success, this is the successor to do znode init
+        echo "Bookkeeper znodes not exist in Zookeeper, do the init to create them."
+        /opt/bookkeeper/bin/bookkeeper shell initnewcluster
+        if [ $? -eq 0 ]; then
+            echo "Bookkeeper znodes init success."
+        else
+            echo "Bookkeeper znodes init failed. please check the reason."
+            exit
+        fi
+    else
+        echo "Other docker instance is doing initialize at the same time, will wait in this instance."
+        tenSeconds=1
+        while [ ${tenSeconds} -lt 10 ]
+        do
+            sleep 10
+            /opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain -server ${BK_zkServers} stat ${BK_zkLedgersRootPath}/available/readonly
+            if [ $? -eq 0 ]; then
+                echo "Waited $tenSeconds * 10 seconds, bookkeeper inited"
```

Comment:
I don't think we need "Waited $tenSeconds * 10 seconds". Here the bookkeeper initialization is already done.

Reply:
Thanks. It is for concurrent initialization. This is not the successor for the init, it should wait the first complete. here it waited and know that initialization is already done, so it will break wait in next line.